File: web/src/data/fixture-source.ts
// FixtureSource — deterministic, offline, and DELIBERATELY SLOW IN THE RIGHT PLACES.
//
// It plays the timeline in fixtures/scripts.ts rather than returning settled answers, so a
// screen built against it has already had to cope with a verdict arriving late and with the
// form growing a field mid-job. That is the entire point of building the surfaces first.
import type {
  Procedure, Job, StepOutcome, Capture, Decision, SealedRecord, Field,
} from "@/generated/types";
import type { DataSource, JobEvent, CaptureInput, BlockedInput, Tier, Unsubscribe } from "./source";
import { procedures } from "./fixtures/procedures";
import { scripts } from "./fixtures/scripts";
import { sealJob, readyToSeal, machineReleased } from "./seal";

let seq = 0;
const id = (p: string) => `${p}_${(++seq).toString(36)}${Date.now().toString(36).slice(-4)}`;
const now = () => new Date().toISOString();

export interface FixtureOptions {
  /** 1 is demo pace. Raise to slow the timeline down for filming; 0 makes it instant for tests. */
  speed?: number;
}

/** The id without its tenant prefix. `{tenant}/{id}` and `{id}` both answer `{id}`. */
const bare = (id: string) => id.slice(id.lastIndexOf("/") + 1);

export class FixtureSource implements DataSource {
  readonly name = "fixture" as const;
  readonly fabricated = true;

  private jobs = new Map<string, Job>();
  private records = new Map<string, SealedRecord>();
  private decisions: Decision[] = [];
  private listeners = new Map<string, Set<(e: JobEvent) => void>>();
  /**
   * Capture id -> something the browser can render it from.
   *
   * The fixture layer has no bucket, so what it holds is the object URL the capture arrived
   * with — the one `CameraLayer` minted from the blob it just grabbed. That URL is alive for
   * as long as the tab is, which is exactly as long as the rest of this store, so the two
   * cannot outlive each other.
   */
  private media = new Map<string, string>();
  private attempts = new Map<string, number>();
  private timers = new Set<ReturnType<typeof setTimeout>>();
  private speed: number;

  constructor(opts: FixtureOptions = {}) {
    this.speed = opts.speed ?? 1;
  }

  /** Cancel every pending beat. Screens call this on unmount so a demo cannot leak timers. */
  dispose() {
    this.timers.forEach(clearTimeout);
    this.timers.clear();
    this.listeners.clear();
  }

  // ---------------------------------------------------------------- reads

  async listProcedures(tenantId: string) {
    return procedures.filter((p) => p.tenant_id === tenantId || tenantId === "*");
  }
  async getProcedure(pid: string) {
    // Bare or `{tenant}/{procedure}`. The fixture layer has one tenant, so the prefix is
    // noise here — but accepting it is what lets a caller scope an id unconditionally
    // instead of branching on which source it happens to be running against.
    const bare = pid.slice(pid.lastIndexOf("/") + 1);
    return procedures.find((p) => p.id === bare) ?? null;
  }
  async getJob(jid: string) {
    // Bare or `{tenant}/{job}`, for the reason `getProcedure` above already gives: a caller
    // should be able to scope an id unconditionally instead of branching on which source it
    // is talking to. LiveSource REQUIRES the scoped form (`splitScoped` returns null for a
    // bare id and it answers null), so a caller written against the interface scopes — and
    // this map is keyed bare, which made every such call miss here and only here.
    return this.jobs.get(jid) ?? this.jobs.get(bare(jid)) ?? null;
  }
  async listJobs(tenantId: string) {
    return [...this.jobs.values()].filter((j) => j.tenant_id === tenantId);
  }
  async getRecord(rid: string) {
    return this.records.get(rid) ?? this.records.get(bare(rid)) ?? null;
  }
  async listRecords(tenantId: string) {
    return [...this.records.values()]
      .filter((r) => tenantId === "*" || this.jobs.get(r.job_id)?.tenant_id === tenantId)
      .sort((a, b) => (a.sealed_at < b.sealed_at ? 1 : -1));
  }
  async listDecisions() {
    return [...this.decisions].sort((a, b) => (a.at < b.at ? 1 : -1));
  }

  // ---------------------------------------------------------------- writes

  async startJob({ procedureId, tenantId, tier }: { procedureId: string; tenantId: string; tier: Tier }) {
    const proc = await this.getProcedure(procedureId);
    if (!proc) throw new Error(`no such procedure: ${procedureId}`);
    const jid = id("job");
    const job: Job = {
      id: jid,
      tenant_id: tenantId,
      procedure_id: proc.id,
      procedure_version: proc.version,
      asset_urn: null,
      technician_id: null,
      status: "draft",
      strictness: proc.strictness,
      tier,
      started_at: now(),
      sealed_at: null,
      steps: proc.steps.map<StepOutcome>((s) => ({
        id: id("out"), job_id: jid, step_id: s.id, status: "pending", fields: [],
      })),
    };
    this.jobs.set(jid, job);
    return job;
  }

  /**
   * The human act that lets the fleet see this job.
   *
   * Present here and not only on LiveSource on purpose: the seam is worth having only while a
   * screen cannot tell the two apart, so a draft gate in one implementation and not the other
   * would be exactly the divergence this interface exists to prevent.
   */
  async finalize(jobId: string): Promise<void> {
    const job = this.jobs.get(jobId);
    if (!job) throw new Error(`no such job: ${jobId}`);
    if (job.status !== "draft") return;
    job.status = "open";
    job.finalized_at = now();
    job.finalized_by = "fixture-technician";
  }

  /** Returns immediately. The verdict arrives later, over subscribe(). */
  async capture(input: CaptureInput): Promise<Capture> {
    const job = this.jobs.get(input.jobId);
    if (!job) throw new Error(`no such job: ${input.jobId}`);
    const cap: Capture = {
      id: id("cap"),
      field_id: `${input.stepId}:${input.fieldKey}`,
      kind: input.kind,
      media_ref: input.mediaRef,
      capture_mode: input.mode,
      capture_surface: input.surface,
      attestation_device_id: input.surface === "browser" ? null : "fixture-device",
      attestation_play_integrity: null,
      redacted: true,
      armor_verdict: "NO_MATCH_FOUND",
      created_at: now(),
    };

    const outcome = job.steps.find((s) => s.step_id === input.stepId);
    if (outcome) {
      const f: Field = {
        id: id("fld"), step_id: input.stepId, key: input.fieldKey,
        kind: input.kind === "audio" ? "text" : input.kind,
        media_ref: cap.id, captured_at: cap.created_at,
        // The class is stamped by the Seal, never here, and never by a model.
        provenance_class: input.surface === "app_instrument" ? "measured" : "inferred",
      };
      // REPLACES rather than appends, because that is what the live source does: `fieldId` in
      // live-source.ts derives the document id from the step and the key, so a retake writes
      // over the answer instead of adding a second one. This pushed unconditionally, which
      // meant a retaken field left TWO entries with the same key — invisible while nothing
      // rendered fields, and two pages of the same slot the moment the handover carousel did.
      // The capture itself is not retracted either way: every attempt is still in `media`, and
      // on the live source every object is still in the bucket.
      const at = outcome.fields.findIndex((x) => x.key === input.fieldKey);
      if (at >= 0) outcome.fields[at] = f; else outcome.fields.push(f);
    }

    // Keyed by the CAPTURE id, because that is what the field points at and therefore what a
    // reader will ask for. `media_ref` on the capture is the object URL; `media_ref` on the
    // field is this id. Two different things, same name — see `mediaUrl` on the seam.
    if (input.mediaRef && input.kind !== "text") this.media.set(cap.id, input.mediaRef);

    this.emit(job.id, { kind: "capture_accepted", stepId: input.stepId, fieldKey: input.fieldKey, at: cap.created_at });
    this.play(job, input.stepId);
    return cap;
  }

  /** What the capture arrived with, or null for a kind that has no object behind it. */
  async mediaUrl(_jobId: string, captureId: string, kind: Field["kind"]): Promise<string | null> {
    if (kind !== "photo" && kind !== "video" && kind !== "scan") return null;
    return this.media.get(captureId) ?? null;
  }

  /** The second exit. There is no skip — this always produces a recorded outcome. */
  async declareBlocked(input: BlockedInput): Promise<StepOutcome> {
    const job = this.jobs.get(input.jobId);
    if (!job) throw new Error(`no such job: ${input.jobId}`);
    const outcome = job.steps.find((s) => s.step_id === input.stepId);
    if (!outcome) throw new Error(`no such step: ${input.stepId}`);

    outcome.reason_kind = input.reasonKind;
    outcome.reason_transcript = input.transcript;
    outcome.reason_audio_ref = input.audioRef ?? null;
    outcome.reason_by = "fixture-technician";
    outcome.reason_at = now();
    outcome.provenance_class = "asserted";

    // Instructor reads the intent and recommends; Foreman disposes. Both are model calls,
    // so both land on the record and both cost money.
    this.after(700, () => {
      outcome.recommendation_text =
        "Nothing here can substitute for it. Record it as deferred and come back once the blocker clears.";
      outcome.recommendation_model = "gemini-3.5-flash";
      this.decide(job, input.stepId, "instructor", "DEFERRED_PROPOSED",
        `Reason read as a blocker, not a refusal: "${input.transcript.slice(0, 80)}"`, "gemini-3.5-flash", 0.00047);
    });
    this.after(1900, () => {
      outcome.status = "deferred";
      outcome.disposition_action = "chase";
      outcome.disposition_at = now();
      this.decide(job, input.stepId, "foreman", "DEFER",
        "Job stays open and the machine stays held. I will check back rather than closing this quietly.", "gemini-3.5-flash", 0.00062);
      this.emit(job.id, { kind: "step_status", stepId: input.stepId, status: "deferred" });
      this.maybeSeal(job);
    });

    return outcome;
  }

  subscribe(jobId: string, cb: (e: JobEvent) => void): Unsubscribe {
    if (!this.listeners.has(jobId)) this.listeners.set(jobId, new Set());
    this.listeners.get(jobId)!.add(cb);
    return () => this.listeners.get(jobId)?.delete(cb);
  }

  // ---------------------------------------------------------------- timeline

  private emit(jobId: string, e: JobEvent) {
    this.listeners.get(jobId)?.forEach((cb) => cb(e));
  }
  private after(ms: number, fn: () => void) {
    if (this.speed === 0) return fn();
    const t = setTimeout(() => { this.timers.delete(t); fn(); }, ms * this.speed);
    this.timers.add(t);
  }
  private decide(job: Job, stepId: string | null, agent: string, verdict: string, rationale: string, model: string | null, cost: number) {
    const d: Decision = {
      id: id("dec"), job_id: job.id, step_id: stepId,
      agent: agent as Decision["agent"], agent_version: `${agent}@1.4.0`,
      model, verdict, rationale, cost_usd: cost, at: now(),
    };
    this.decisions.push(d);
    this.emit(job.id, { kind: "decision", stepId, decision: d });
  }

  private play(job: Job, stepId: string) {
    const script = scripts[job.procedure_id]?.[stepId];
    if (!script) return;
    const key = `${job.id}:${stepId}`;
    const attempt = this.attempts.get(key) ?? 0;
    this.attempts.set(key, attempt + 1);
    const beats = script[Math.min(attempt, script.length - 1)];

    for (const beat of beats) {
      this.after(beat.at, () => {
        if (beat.kind === "decision") {
          this.decide(job, stepId, beat.agent, beat.verdict, beat.rationale, beat.model ?? null, beat.cost ?? 0);
        } else if (beat.kind === "add_field") {
          this.emit(job.id, { kind: "add_field", stepId, field: beat.field });
        } else if (beat.kind === "escalate") {
          this.emit(job.id, { kind: "escalated", stepId, question: beat.question });
        } else {
          const o = job.steps.find((s) => s.step_id === stepId);
          if (o) o.status = beat.status;
          this.emit(job.id, { kind: "step_status", stepId, status: beat.status });
          this.maybeSeal(job);
        }
      });
    }
  }

  private maybeSeal(job: Job) {
    // The fixture runs the SAME seal logic as the server, which is the whole point of
    // data/seal.ts — so it owes `readyToSeal` the same procedure the server owes it, or an
    // optional step would hold a fixture job open while sealing the live one.
    const procedure = procedures.find((p) => p.id === job.procedure_id) ?? null;
    if (job.status === "sealed" || !readyToSeal(job, procedure)) return;
    job.status = "sealed";
    job.sealed_at = now();
    const rec = sealJob(job, this.decisions.filter((d) => d.job_id === job.id), { public: job.tenant_id === "anon" });
    this.records.set(rec.id, rec);
    if (!machineReleased(job, procedure)) this.emit(job.id, { kind: "held", reason: "a step was explained rather than performed" });
    this.emit(job.id, { kind: "sealed", recordId: rec.id });
  }
}

File: android/app/src/main/java/ink/warrant/data/FixtureSource.kt
package ink.warrant.data

import ink.warrant.contract.Agent
import ink.warrant.contract.Capture
import ink.warrant.contract.CaptureKind
import ink.warrant.contract.Decision
import ink.warrant.contract.Field
import ink.warrant.contract.FieldKind
import ink.warrant.contract.Job
import ink.warrant.contract.JobStatus
import ink.warrant.contract.Procedure
import ink.warrant.contract.ProvenanceClass
import ink.warrant.contract.Reading
import ink.warrant.contract.SealedRecord
import ink.warrant.contract.StepOutcome
import ink.warrant.contract.StepStatus
import ink.warrant.contract.Tier
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import java.util.concurrent.atomic.AtomicLong

/**
 * FixtureSource — deterministic, offline, and DELIBERATELY SLOW IN THE RIGHT PLACES.
 *
 * It plays the timeline in [scripts] rather than returning settled answers, so a screen built
 * against it has already had to cope with a verdict arriving late and with the form growing a
 * field mid-job. That is the entire point of building the surfaces before the backend: when
 * LiveSource replaces this, nothing about the screens has to change, because they were never
 * written in a world where verification was instant.
 *
 * Ported from `web/src/data/fixture-source.ts`. State is guarded by a [Mutex] rather than
 * relying on a single-threaded event loop, which is the one real difference from the original.
 */
class FixtureSource(
    /**
     * 1.0 is demo pace. Raise it to slow the timeline down for filming; 0.0 makes every beat
     * fire immediately, which is what the tests use.
     */
    private val speed: Double = 1.0,
    parent: CoroutineScope? = null,
) : DataSource {

    override val name = "fixture"
    override val fabricated = true

    private val scope = parent ?: CoroutineScope(SupervisorJob())
    private val mutex = Mutex()

    private val jobs = mutableMapOf<String, Job>()
    /**
     * Kept for the same reason the live path keeps a `captures` subcollection: a field points
     * at a capture, and the capture is what knows where the bytes are. Without this the
     * fixture could store a photograph and then have no way back to it — which is exactly the
     * gap that made records render "evidence stored, not reachable from here".
     */
    private val captures = mutableMapOf<String, Capture>()
    private val records = mutableMapOf<String, SealedRecord>()
    private val decisions = mutableListOf<Decision>()
    private val attempts = mutableMapOf<String, Int>()
    private val streams = mutableMapOf<String, MutableSharedFlow<JobEvent>>()

    private val seq = AtomicLong(0)
    private fun id(prefix: String) = "${prefix}_${seq.incrementAndGet().toString(36)}"

    /** Cancel every pending beat. A demo must not leak timers between runs. */
    fun dispose() {
        if (parentWasNull) scope.cancel()
    }

    private val parentWasNull = parent == null

    // ---------------------------------------------------------------- reads

    override suspend fun listProcedures(tenantId: String): List<Procedure> =
        procedures.filter { it.tenantId == tenantId || tenantId == "*" }

    override suspend fun getProcedure(id: String): Procedure? =
        procedures.firstOrNull { it.id == id }

    override suspend fun getJob(id: String): Job? = mutex.withLock { jobs[id] }

    override suspend fun listJobs(tenantId: String): List<Job> =
        mutex.withLock {
            jobs.values
                .filter { it.tenantId == tenantId || tenantId == "*" }
                .sortedByDescending { it.startedAt }
        }

    /**
     * Gone from the map, and gone from the stream.
     *
     * Dropping the flow matters as much as dropping the job. The fixture's whole point is that
     * verdicts arrive on a timer after the technician has walked away, so a deleted job leaves
     * beats already scheduled against it. They all re-read `jobs[jobId]` and fall out when it
     * is missing — but a subscriber still holding the old flow would sit there forever waiting
     * for events that can no longer be produced, which on screen looks exactly like a job that
     * is thinking rather than one that is gone.
     */
    override suspend fun deleteJob(id: String) {
        mutex.withLock {
            // Already gone is a success. See the note on DataSource.deleteJob.
            val job = jobs[id] ?: return@withLock
            require(job.status != JobStatus.SEALED) { "a sealed job cannot be deleted: $id" }
            jobs.remove(id)
            streams.remove(id)
        }
    }

    /**
     * The same two hops the live path makes — capture id to capture, capture to bytes — with
     * the second landing on this device instead of in Cloud Storage. There is nothing to
     * exchange for a signed URL here, only the check that the file is really there.
     *
     * A capture that has been cleaned up, or one this fixture never saw, is null, exactly as a
     * missing object is on the live path. The screens cannot tell the difference, which is the
     * test that the seam is in the right place.
     */
    override suspend fun mediaUrl(jobId: String, captureId: String, kind: FieldKind): String? {
        // A kind with no object behind it. Same answer as live, for the same reason.
        if (Media.extension(kind) == null) return null
        val capture = mutex.withLock { captures[captureId] } ?: return null
        return capture.mediaRef.takeIf { it.isNotBlank() && java.io.File(it).exists() }
    }

    override suspend fun getRecord(id: String): SealedRecord? = mutex.withLock { records[id] }

    override suspend fun listRecords(tenantId: String): List<SealedRecord> =
        mutex.withLock {
            records.values
                .filter { it.tenantId == tenantId || tenantId == "*" }
                .sortedByDescending { it.sealedAt }
        }

    override suspend fun listDecisions(tenantId: String): List<Decision> =
        mutex.withLock { decisions.sortedByDescending { it.at } }

    // ---------------------------------------------------------------- writes

    override suspend fun startJob(procedureId: String, tenantId: String, tier: Tier): Job {
        val proc = getProcedure(procedureId)
            ?: throw IllegalArgumentException("no such procedure: $procedureId")

        // Refused, never downgraded. A procedure that needs an instrument does not quietly
        // become a photo-only job because the surface could not manage it.
        require(surfaceCanRun(proc, tier)) {
            "${proc.key} needs the ${proc.minimumTier} tier; this surface offers $tier"
        }

        val jid = id("job")
        val job = Job(
            id = jid,
            tenantId = tenantId,
            procedureId = proc.id,
            procedureVersion = proc.version,
            assetUrn = null,
            technicianId = null,
            status = JobStatus.OPEN,
            strictness = proc.strictness,
            tier = tier,
            startedAt = nowIso(),
            sealedAt = null,
            steps = proc.steps.map { step ->
                StepOutcome(
                    id = id("out"), jobId = jid, stepId = step.id,
                    status = StepStatus.PENDING, fields = emptyList(),
                )
            },
        )
        mutex.withLock { jobs[jid] = job }
        return job
    }

    /** Returns as soon as the evidence is stored. The verdict arrives later, over [subscribe]. */
    override suspend fun capture(input: CaptureInput): Capture {
        val cap = Capture(
            id = id("cap"),
            fieldId = "${input.stepId}:${input.fieldKey}",
            kind = input.kind,
            mediaRef = input.mediaRef,
            captureMode = input.mode,
            captureSurface = input.surface,
            attestationDeviceId = "fixture-device",
            attestationPlayIntegrity = null,
            redacted = input.redacted,
            armorVerdict = "NO_MATCH_FOUND",
            createdAt = nowIso(),
        )

        val job = mutex.withLock {
            captures[cap.id] = cap
            val j = jobs[input.jobId] ?: throw IllegalArgumentException("no such job: ${input.jobId}")
            val field = Field(
                id = id("fld"),
                stepId = input.stepId,
                key = input.fieldKey,
                kind = when (input.kind) {
                    CaptureKind.PHOTO -> FieldKind.PHOTO
                    CaptureKind.VIDEO -> FieldKind.VIDEO
                    CaptureKind.SCAN -> FieldKind.SCAN
                    CaptureKind.AUDIO, CaptureKind.TEXT -> FieldKind.TEXT
                },
                // The same split the live source makes: a text answer has no object, and
                // `isFilled` reads valueText for this kind — a typed answer left only in
                // mediaRef reads as an empty field on the fixtures the demo runs on.
                mediaRef = if (input.kind.hasObject) cap.id else null,
                valueText = if (input.kind.hasObject) null else input.mediaRef,
                capturedAt = cap.createdAt,
                // The class is stamped by the Seal, never here, and never by a model. What is
                // recorded now is only how the evidence was made.
                provenanceClass = null,
            )
            val updated = j.copy(
                steps = j.steps.map { o ->
                    if (o.stepId == input.stepId) o.copy(fields = o.fields + field) else o
                },
            )
            jobs[j.id] = updated
            updated
        }

        emit(job.id, JobEvent.CaptureAccepted(input.stepId, input.fieldKey, cap.createdAt))
        play(job.id, input.stepId)
        return cap
    }

    /**
     * A number from a paired instrument. There is no path here for a typed value — that is not
     * an oversight, it is the whole distinction between measured and asserted.
     */
    override suspend fun submitReading(input: ReadingInput): Reading {
        val reading = Reading(
            id = id("rdg"),
            fieldId = "${input.stepId}:${input.fieldKey}",
            componentId = null,
            key = input.fieldKey,
            value = input.value,
            unit = input.unit,
            toolId = input.toolId,
            at = nowIso(),
        )

        val job = mutex.withLock {
            val j = jobs[input.jobId] ?: throw IllegalArgumentException("no such job: ${input.jobId}")
            val field = Field(
                id = id("fld"),
                stepId = input.stepId,
                key = input.fieldKey,
                kind = FieldKind.MEASUREMENT,
                valueNumber = input.value,
                unit = input.unit,
                toolId = input.toolId,
                capturedAt = reading.at,
                provenanceClass = null,
            )
            val updated = j.copy(
                steps = j.steps.map { o ->
                    if (o.stepId == input.stepId) o.copy(fields = o.fields + field) else o
                },
            )
            jobs[j.id] = updated
            updated
        }

        emit(job.id, JobEvent.ReadingArrived(input.stepId, input.fieldKey, reading))
        play(job.id, input.stepId)
        return reading
    }

    /** The second exit. There is no skip — this always produces a recorded outcome. */
    override suspend fun declareBlocked(input: BlockedInput): StepOutcome {
        val outcome = mutex.withLock {
            val j = jobs[input.jobId] ?: throw IllegalArgumentException("no such job: ${input.jobId}")
            val o = j.steps.firstOrNull { it.stepId == input.stepId }
                ?: throw IllegalArgumentException("no such step: ${input.stepId}")
            val updated = o.copy(
                reasonKind = input.reasonKind,
                reasonTranscript = input.transcript,
                reasonAudioRef = input.audioRef,
                reasonBy = "fixture-technician",
                reasonAt = nowIso(),
                // A stated reason is always asserted — a named human said it, at this time.
                provenanceClass = ProvenanceClass.ASSERTED,
            )
            jobs[j.id] = j.copy(steps = j.steps.map { if (it.stepId == input.stepId) updated else it })
            updated
        }

        // The Instructor reads the intent and recommends; the Foreman disposes. Both are model
        // calls, so both land on the record and both cost money.
        scope.launch {
            beat(700)
            mutex.withLock {
                val j = jobs[input.jobId] ?: return@withLock
                jobs[j.id] = j.copy(
                    steps = j.steps.map {
                        if (it.stepId == input.stepId) {
                            it.copy(
                                recommendationText = "Nothing here can substitute for it. Record " +
                                    "it as deferred and come back once the blocker clears.",
                                recommendationModel = "gemini-3.5-flash",
                            )
                        } else {
                            it
                        }
                    },
                )
            }
            decide(
                input.jobId, input.stepId, Agent.INSTRUCTOR, "DEFERRED_PROPOSED",
                "Reason read as a blocker, not a refusal: \"${input.transcript.take(80)}\"",
                "gemini-3.5-flash", 0.00047,
            )

            beat(1200)
            mutex.withLock {
                val j = jobs[input.jobId] ?: return@withLock
                jobs[j.id] = j.copy(
                    steps = j.steps.map {
                        if (it.stepId == input.stepId) {
                            it.copy(
                                status = StepStatus.DEFERRED,
                                dispositionAction = ink.warrant.contract.DispositionAction.CHASE,
                                dispositionAt = nowIso(),
                            )
                        } else {
                            it
                        }
                    },
                )
            }
            decide(
                input.jobId, input.stepId, Agent.FOREMAN, "DEFER",
                "Job stays open and the machine stays held. I will check back rather than " +
                    "closing this quietly.",
                "gemini-3.5-flash", 0.00062,
            )
            emit(input.jobId, JobEvent.StepStatusChanged(input.stepId, StepStatus.DEFERRED))
            maybeSeal(input.jobId)
        }

        return outcome
    }

    /**
     * An answer to a question an agent asked. Beside the question, never over it.
     *
     * The step is left PENDING on purpose. The person has spoken; the fleet has not ruled yet,
     * and a screen that flipped the step to performed the moment somebody typed would be the
     * tick in the box this product exists to replace. The Skeptic below is what settles it,
     * and it is allowed to disagree.
     */
    override suspend fun respond(input: ResponseInput): StepOutcome {
        val at = nowIso()
        val outcome = mutex.withLock {
            val j = jobs[input.jobId] ?: throw IllegalArgumentException("no such job: ${input.jobId}")
            val o = j.steps.firstOrNull { it.stepId == input.stepId }
                ?: throw IllegalArgumentException("no such step: ${input.stepId}")
            val updated = o.copy(
                escalationAnswer = input.answer,
                escalationAnsweredBy = input.by,
                escalationAnsweredAt = at,
                // A stated answer is asserted, exactly like a stated reason: a named human
                // said it, at this time, and nothing checked it.
                provenanceClass = ProvenanceClass.ASSERTED,
                // The fleet could not act, and now a person has spoken. Clearing this is what
                // takes the step off the "stuck" list without claiming it passed.
                holdReason = null,
            )
            jobs[j.id] = j.copy(steps = j.steps.map { if (it.stepId == input.stepId) updated else it })
            updated
        }

        scope.launch {
            beat(900)
            decide(
                input.jobId, input.stepId, Agent.SKEPTIC, "ANSWER_ACCEPTED",
                "The answer names the thing I could not see in the evidence: " +
                    "\"${input.answer.take(80)}\". Asserted, not measured — I am recording " +
                    "whose word it is.",
                "gemini-3.5-flash", 0.00051,
            )

            beat(1100)
            mutex.withLock {
                val j = jobs[input.jobId] ?: return@withLock
                jobs[j.id] = j.copy(
                    steps = j.steps.map {
                        if (it.stepId == input.stepId) {
                            it.copy(status = StepStatus.PERFORMED, adjudicatedAt = nowIso())
                        } else {
                            it
                        }
                    },
                )
            }
            emit(input.jobId, JobEvent.StepStatusChanged(input.stepId, StepStatus.PERFORMED))
            maybeSeal(input.jobId)
        }

        return outcome
    }

    override fun subscribe(jobId: String): Flow<JobEvent> = streamFor(jobId).asSharedFlow()

    // ---------------------------------------------------------------- timeline

    @Synchronized
    private fun streamFor(jobId: String): MutableSharedFlow<JobEvent> =
        streams.getOrPut(jobId) {
            // replay = 0: a late subscriber does not get a burst of history. extraBufferCapacity
            // keeps emit() from suspending when nobody is listening yet.
            MutableSharedFlow(replay = 0, extraBufferCapacity = 64)
        }

    private suspend fun emit(jobId: String, event: JobEvent) {
        streamFor(jobId).emit(event)
    }

    /** One beat of the demo clock. speed 0 fires everything immediately, which tests rely on. */
    private suspend fun beat(ms: Long) {
        if (speed > 0.0) delay((ms * speed).toLong())
    }

    private suspend fun decide(
        jobId: String,
        stepId: String?,
        agent: Agent,
        verdict: String,
        rationale: String,
        model: String?,
        cost: Double,
    ) {
        val d = Decision(
            id = id("dec"), jobId = jobId, stepId = stepId, agent = agent,
            agentVersion = "${agent.name.lowercase()}@1.4.0",
            model = model, verdict = verdict, rationale = rationale,
            costUsd = cost, at = nowIso(),
        )
        mutex.withLock { decisions += d }
        emit(jobId, JobEvent.DecisionMade(stepId, d))
    }

    /**
     * Plays the script for one step. The attempt counter is what makes ADD FIELD real: the
     * first capture on a step runs beat list 0, the second runs list 1, and a step whose script
     * has one entry simply replays it.
     */
    private fun play(jobId: String, stepId: String) {
        scope.launch {
            val job = mutex.withLock { jobs[jobId] } ?: return@launch
            val stepScript = scripts[job.procedureId]?.get(stepId) ?: return@launch
            val key = "$jobId:$stepId"
            val attempt = mutex.withLock {
                val a = attempts[key] ?: 0
                attempts[key] = a + 1
                a
            }
            val beats = stepScript[minOf(attempt, stepScript.size - 1)]

            var elapsed = 0L
            for (b in beats) {
                beat(b.at - elapsed)
                elapsed = b.at
                when (b) {
                    is DemoBeat.Decide ->
                        decide(jobId, stepId, b.agent, b.verdict, b.rationale, b.model, b.cost)

                    is DemoBeat.AddField ->
                        emit(jobId, JobEvent.FieldAdded(stepId, b.field))

                    is DemoBeat.Escalate ->
                        emit(jobId, JobEvent.Escalated(stepId, b.question))

                    is DemoBeat.Status -> {
                        mutex.withLock {
                            val j = jobs[jobId] ?: return@withLock
                            jobs[jobId] = j.copy(
                                steps = j.steps.map {
                                    if (it.stepId == stepId) it.copy(status = b.status) else it
                                },
                            )
                        }
                        emit(jobId, JobEvent.StepStatusChanged(stepId, b.status))
                        maybeSeal(jobId)
                    }
                }
            }
        }
    }

    private suspend fun maybeSeal(jobId: String) {
        val (job, record) = mutex.withLock {
            val j = jobs[jobId] ?: return
            if (j.status == JobStatus.SEALED || !readyToSeal(j)) return
            val sealedJob = j.copy(status = JobStatus.SEALED, sealedAt = nowIso())
            jobs[jobId] = sealedJob
            val rec = sealJob(
                sealedJob,
                decisions.filter { it.jobId == jobId },
                public = sealedJob.tenantId == "anon",
            )
            records[rec.id] = rec
            sealedJob to rec
        }

        if (!machineReleased(job)) {
            emit(jobId, JobEvent.Held("a step was explained rather than performed"))
        }
        emit(jobId, JobEvent.Sealed(record.id))
    }
}
